tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+                        "num_rows INT NOT NULL,"\
+                        "num_cols INT NOT NULL,"\
+                        "PRIMARY KEY (video, id));";
+
+    char *error = nullptr;
+    int result = sqlite3_exec(db_, createTable, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating layouts table: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+
+    const char *createIndex = "CREATE INDEX layout_index ON layouts (video, first_frame, last_frame)";
+    result = sqlite3_exec(db_, createIndex, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating layouts index: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+}
+
+void LayoutDatabase::createWidthsTable() {
+    const char *createWidthTable = "CREATE TABLE widths ("\
+                            "video TEXT NOT NULL,"\
+                            "id INT NOT NULL,"\
+                            "width INT NOT NULL,"\
+                            "position INT NOT NULL,"\
+                            "FOREIGN KEY (video, id) REFERENCES layouts(video, id))";
+    char *error = nullptr;
+    int result = sqlite3_exec(db_, createWidthTable, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating layout width table: " << sqlite3_errmsg(db_) << std::endl;;
+        sqlite3_free(error);
+    }
+
+    const char *createIndex = "CREATE INDEX widths_index ON widths(video, id)";
+    result = sqlite3_exec(db_, createIndex, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating widths index: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+}
+
+void LayoutDatabase::createHeightsTable() {
+    const char *createHeightTable = "CREATE TABLE heights ("\
+                            "video TEXT NOT NULL,"\
+                            "id INT NOT NULL,"\
+                            "height INT NOT NULL,"\
+                            "position INT NOT NULL,"\
+                            "FOREIGN KEY (video, id) REFERENCES layouts(video, id))";
+    char *error = nullptr;
+    int result = sqlite3_exec(db_, createHeightTable, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating layout height table: " << sqlite3_errmsg(db_) << std::endl;;
+        sqlite3_free(error);
+    }
+
+    const char *createIndex = "CREATE INDEX heights_index ON heights(video, id)";
+    result = sqlite3_exec(db_, createIndex, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating heights index: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+}
+
+std::vector<int> LayoutDatabase::tileLayoutIdsForFrame(const std::string &video, unsigned int frameNumber) const {
+    std::string selectLayoutIds = "SELECT id FROM layouts WHERE video = ? AND ? BETWEEN first_frame AND last_frame";
```

Comment:
Because we are presumably going to be executing this query multiple times, it may make sense initialize it as a prepared statement, and then do bind/reset for each invocation. An example of this is in `SemanticIndexSQLite::initializeStatements(), ::destroyStatements(), ::addMetadata()`

(Ditto for the rest of the queries in this class)